src/Engine/Physics/TriggerObserver.cpp

```diff
@@ -0,0 +1,112 @@
+        return &rigidBody;
+    }
+
+    TriggerObserver::IntersectionPhase TriggerObserver::GetPhase() const {
+        return phase;
+    }
+
+    void TriggerObserver::PostIntersectionTest() {
+        if (didCallback) {
+            // Intersection happened; update phase accordingly.
+            switch (phase) {
+                case IntersectionPhase::Enter: {
+                    phase = IntersectionPhase::Retained;
+                    break;
+                }
+                case IntersectionPhase::Leave: {
+                    phase = IntersectionPhase::Enter;
+                    break;
+                }
+                case IntersectionPhase::None: {
+                    phase = IntersectionPhase::Enter;
+                    break;
+                }
+                default: {
+                    assert(phase == IntersectionPhase::Retained);
+                }
+            }
+        } else {
+            // No intersection happened; update phase accordingly.
+            switch (phase) {
+                case IntersectionPhase::Enter: {
+                    phase = IntersectionPhase::Leave;
+                    break;
+                }
+                case IntersectionPhase::Retained: {
+                    phase = IntersectionPhase::Leave;
+                    break;
+                }
+                case IntersectionPhase::Leave: {
+                    phase = IntersectionPhase::None;
+                    break;
+                }
+                default: {
+                    assert(phase == IntersectionPhase::None);
+                }
+            }
+        }
+
+        // Call event handler accordingly.
+        switch (phase) {
+            case IntersectionPhase::Enter: {
+                if (enterHandler) {
+                    enterHandler();
+                }
+                break;
+            }
+            case IntersectionPhase::Retained: {
+                if (retainHandler) {
+                    retainHandler();
+                }
+                break;
+            }
+            case IntersectionPhase::Leave: {
+                if (leaveHandler) {
+                    leaveHandler();
+                }
+                break;
+            }
+            default: {
+                assert(phase == IntersectionPhase::None);
+            }
+        }
+
+        // Prepare for next invocation.
+        didCallback = false;
+    }
+
+    void TriggerObserver::OnEnter(const std::function<void()>& handler) {
+        enterHandler = handler;
+    }
+
+    void TriggerObserver::OnRetain(const std::function<void()>& handler) {
+        retainHandler = handler;
+    }
+
+    void TriggerObserver::OnLeave(const std::function<void()>& handler) {
+        leaveHandler = handler;
+    }
+
+    // Called with each contact for our own processing. This is where we can
+    // provide custom intersection handling.
+    btScalar TriggerObserver::addSingleResult(btManifoldPoint& cp,
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') The function 'addSingleResult' is never used. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)